### PR TITLE
Implement Jenkins#updateNode()

### PR DESCRIPTION
**Pull request by felfert**
Hi,

This trivial PR adds Jenkins#updateNode(). I propose this in order to prevent ugly workarounds in [this PR](https://github.com/OpenDataSpace/jclouds-plugin/commit/5814b4db03cdd97f3b3183133730e1d57d73b917) for the jclouds-plugin.

Cheers
 -Fritz

### CODE REVIEW

File: core/src/main/java/jenkins/model/Nodes.java [hunk 1]
```diff
@@ -149,6 +149,25 @@ public void run() {
...
+    public void updateNode(final @Nonnull Node node) throws IOException {
+        if (node == nodes.get(node.getNodeName())) {
+            // no need for a full save() so we just do the minimum
+            if (node instanceof EphemeralNode) {
+                Util.deleteRecursive(new File(getNodesDir(), node.getNodeName()));
+            } else {
+                XmlFile xmlFile = new XmlFile(Jenkins.XSTREAM,
+                        new File(new File(getNodesDir(), node.getNodeName()), "config.xml"));
+                xmlFile.write(node);
+            }
```
Comment (jtnord) [inline core/src/main/java/jenkins/model/Nodes.java]: this should fire a saveableListener as its state has been changed.

Comment (felfert) [inline core/src/main/java/jenkins/model/Nodes.java]: Ok, dedup'd the code. I'm not shure about firing saveableListener.

File: core/src/main/java/jenkins/model/Nodes.java [hunk 2]
```diff
@@ -137,14 +137,36 @@ public void run() {
...
+
+    /**
+     * Updates an existing node on disk.
+     *
+     * @param node the node to be updated.
+     * @throws IOException if the node could not be persisted.
+     */
+    public void updateNode(final @Nonnull Node node) throws IOException {
+        if (node == nodes.get(node.getNodeName())) {
+            persistNode(node);
```
Comment (jtnord) [inline core/src/main/java/jenkins/model/Nodes.java]: sorry didn't notice this earlier - but this should at least log if a call was made and the `if` statement results to false.
I'm wondering if this should be more than a log as well and actually throw an exception.

**Comment by jtnord** [comment]
@stephenc any reason not to fire the `saveable` listener here that you know of (as the `removeNode` and `addNode` don't which seems strange).

quite surprised I did not see any existing tests of manipulating nodes (except the cli)

@felfert I think you also need to update the Jenkins labels list with a call to `jenkins.trimLabels()`

**Comment by felfert** [comment]
@jtnord About logging (or even throwing) if the node instance is not in the list already: Modeling it after addNode/removeNode: Both silently ignore instances that dont apply. Updated the JavaDoc accordingly.

Also: Added the jenkins.trimLabels() call - thanks for the hint.

**Comment by stephenc** [comment]
I think it was just oversight that I missed the call to Saveable listener... we should probably fix that by adding the call to the `persistNode` method with this PR

**Comment by felfert** [comment]
@stephenc Done.

File: core/src/main/java/jenkins/model/Nodes.java [hunk 3]
```diff
@@ -137,14 +137,44 @@ public void run() {
...
+    }
+
+    /**
+     * Updates an existing node on disk. If the node instance is not in the list of nodes, then this will be a no-op, even if
+     * there is another instance with the same {@link Node#getNodeName()}.
+     *
+     * @param node the node to be updated.
+     * @throws IOException if the node could not be persisted.
+     */
+    public void updateNode(final @Nonnull Node node) throws IOException {
```
Comment (stephenc) [inline core/src/main/java/jenkins/model/Nodes.java]: I would recommend returning a boolean, `true` for a successful update and `false` for the "no-op" case

File: core/src/main/java/jenkins/model/Jenkins.java [hunk 4]
```diff
@@ -1735,6 +1735,13 @@ public void removeNode(@Nonnull Node n) throws IOException {
         nodes.removeNode(n);
     }
 
+    /**
+     * Updates an existing {@link Node} on disk.
+     */
+    public void updateNode(Node n) throws IOException {
```
Comment (stephenc) [inline core/src/main/java/jenkins/model/Jenkins.java]: return a boolean esp. given the semantics of the update operation if there is an existing node of the same name that is not the provided node instance

**Pull request merged**